Remove event model - phase1

gonzaponte opened the pull request:
This PR is the first step towards removing the superfluous parts of the event model. It replaces the `Hit` class and its subsidiaries with DataFrames, which are much more convenient.

gonzaponte committed:
Rewrite `sipms_as_hits` to produce hits DataFrames

gonzaponte committed:
Rewrite `count_valid_hits` to work with DataFrames

gonzaponte committed:
Rewrite `hits_merger` component to work with DataFrames

gonzaponte committed:
Rewrite `hits_functions` in terms of DataFrames

gonzaponte committed:
Rewrite `hits_functions_test` in terms of DataFrames

gonzaponte committed:
Rewrite `hits_corrector` to work with DataFrames

gonzaponte committed:
Rewrite `hits_writer` to work with DataFrames

gonzaponte committed:
Rewrite `hits_and_kdst_from_files` to provide DataFrames

gonzaponte committed:
Rewrite `hits_thresholder` to work with DataFrames

gonzaponte committed:
Rewrite Beersheba to work with DataFrames

gonzaponte committed:
Rewrite `hit_dropper` to work with DataFrames

gonzaponte committed:
Rewrite Esmeralda to work with DataFrames

gonzaponte committed:
Rewrite `hit_builder` to work with DataFrames

gonzaponte committed:
Rewrite Penthesilea to work with DataFrames

gonzaponte committed:
Rewrite Isaura to work with DataFrames

gonzaponte committed:
Refactor `compute_and_write_track_info`

gonzaponte committed:
Add TODO

gonzaponte committed:
Rewrite `test_sophronia_keeps_hitless_events`

jwaiton commented on invisible_cities/cities/components.py:
```diff
@@ -598,17 +598,17 @@ def hits_and_kdst_from_files( paths : List[str]
                 event_number, timestamp = evtinfo.fetch_all_fields()
-                hits = hits_from_df(hits_df.loc[hits_df.event == event_number])
-                if len(hits):
-                    hits = hits[event_number]
-                else:
+                this_event = lambda df: df.event == event_number
+                hits = hits_df.loc[this_event]
+                kdst = kdst_df.loc[this_event]
```
Could this not be written in a manner that doesn't require multiple iterative lookups? Perhaps instead with `kdst.groupby('event')`? Unless this isn't possible due to events that have no kdst and no hits.

Of course, then you need to fetch the `evtinfo` anyway, so perhaps theres no neat way around this outside of sorting the hits, kdst and evtinfo then grouping that way.

gonzaponte replied:
I don't recall the details but I think I had a problem with groupby. Indeed `groupby` combined with `get_group` is faster. But your own explanation sounds like something I could have run into. If either `hits_df` or `kdst_df` doesn't have an event it will raise an error. The speedup is not significant in this context so I'm tempted to leave it as is, since it is very self-explanatory.

jwaiton commented on invisible_cities/cities/components.py:
```diff
@@ -949,24 +954,24 @@ def build_hits(pmap, selector_output, event_number, timestamp):
-                z_slice = (t_slice - s1_t) * units.ns * drift_v
-                e_slice = peak.pmts.sum_over_sensors[slice_no]
+            slice_zs = (peak.times - s1_t) * units.ns * drift_v
+            slice_es = peak.pmts.sum_over_sensors
```
Perhaps this is my ignorance, but this area of the code (functions related to `SensorResponses()` such as `sum_over_sensors` as an example in this case) is quite dark.

I believe I understand how it works, but it may be useful to improve the docstrings in tandem while the functions are being improved.

gonzaponte replied:
Which docstrings are you referring to?
Can you ellaborate on what you find dark so I can focus the docstring on that?

jwaiton replied:
The function `hits_and_kdst_from_file()` itself doesn't have a great docstring.

I'll follow through my process to hopefully clarify what I find dark about this section of the code. Sorry if this is a bit tedious:
- `peak` uses class components like `times` and `pmts`. I wasn't sure where these came from or how they were defined, and my IDE couldn't pick up their origin.
- Tracing back through the function, I find that peak comes from `selector_output` as seen [here](https://github.com/gonzaponte/IC/blob/378b6425438f8dbd5acef1a0ad0b8385405b6190/invisible_cities/cities/components.py#L948) in the `build_hits()` function. 
- Neither `build_hits()` nor `hit_builder()` have type formatting or a docstring explaining what the `selector_output` is, so I have to find it being used in [`penthesilea`](https://github.com/gonzaponte/IC/blob/master/invisible_cities/cities/penthesilea.py#L95).
- From here, I can see that the `selector_output` is from [`peak_classifier()`](https://github.com/gonzaponte/IC/blob/master/invisible_cities/cities/components.py#L811), but you have to understand how the dataflow system works to see this. `peak_classifier()` is a `partial` of the [`S12Selector` class](https://github.com/gonzaponte/IC/blob/master/invisible_cities/filters/s1s2_filter.py#L66).
- The `S12Selector` class itself has no overarching docstring explaining what it does, although I feel it's relatively straightforward. It uses the [`_Peak` class](https://github.com/gonzaponte/IC/blob/master/invisible_cities/evm/pmaps.py#L27) which itself holds the `times` and `pmts` information, although it also has no docstrings.

I feel like the majority of these steps could be avoided if `hit_builder()` had type formatting and a docstring, but if that isn't the case, the path back to the `_Peak` class is not that simple/is relatively undocumented. I don't think altering this is within the scope of this PR though :)

gonzaponte replied:
> `peak` uses class components like times and pmts. I wasn't sure where these came from or how they were defined, and my IDE couldn't pick up their origin.

These are in the constructor. Perhaps it doesn't peak them up because it derives from a base class?

I am adding some documentation for the other bits.

jwaiton commented on invisible_cities/cities/components.py:
```diff
@@ -929,6 +929,11 @@ def sipm_positions(dbfile, run_number):
 
+def e_from_q(qs, e_slice):
+    epsilon = np.finfo(np.float64).eps
+    return qs * e_slice / (qs.sum() + epsilon)
```
This replaces [split_energy()](https://github.com/gonzaponte/IC/blob/378b6425438f8dbd5acef1a0ad0b8385405b6190/invisible_cities/reco/hits_functions.py#L9), but perhaps it shouldn't sit like this in `components.py`. Would there be a suitable file for it to sit in?

gonzaponte replied:
Agreed. Probably in `hits_functions`.

jwaiton commented on invisible_cities/cities/components.py:
```diff
@@ -949,24 +954,24 @@ def build_hits(pmap, selector_output, event_number, timestamp):
+                    clusters = slice_reco(xys, sipm_qs)
+                    qs       = np.array([c.Q for c in clusters])
+                    es       = e_from_q(qs, e_slice)
                     for c, e in zip(clusters, es):
                         hit  = Hit(peak_no, c, z_slice, e, xy_peak)
                         hitc.hits.append(hit)
```
The `hit` class is retained here, and used for `penthesilea`. Is this intended?

jwaiton replied:
I have a feeling that the `hit_builder()` and `sipm_as_hits()` function very similarly (returns hits from similar inputs). Is there a reason these two functions aren't merged?

jwaiton replied:
Upon further reading, it appears that `penthesilea` is no longer used in the official production flow. Is there any need to update these functions rather than deprecate/remove them?

gonzaponte replied:
Yes, `sipms_as_hits` is essentially a rewrite of `hit_builder` to make it do what it was already doing with a less convoluted approach. The latter is more general, and could reproduce the former under a specific (magical) configuration.

The idea behind updating these functions is to keep `penthesilea` alive as we didn't (and still don't) know what reconstruction strategy will be necessary in this new detector.

jwaiton commented on invisible_cities/cities/components.py:
```diff
@@ -1017,26 +1016,35 @@ def build_hits(pmap, selector_output, event_number, timestamp):
+                sipm_xy  =              xys[over_thr]
+                sipm_qs  =          sipm_qs[over_thr] if nonempty else [NN]
+                sipm_xs  =              sipm_xy[:, 0] if nonempty else [NN]
+                sipm_ys  =              sipm_xy[:, 1] if nonempty else [NN]
```
The behaviour for `!nonempty` providing a `NN` result (`-9999`), is a novel component of this function. Should it be tested that this works as expected/is accounted for appropriately?

gonzaponte replied:
Yes.

jwaiton commented on invisible_cities/cities/sophronia.py:
```diff
@@ -70,8 +71,8 @@
 @check_annotations
-def count_valid_hits(hitc : HitCollection):
-    return sum(1 for hit in hitc.hits if hit.Q != NN)
+def count_valid_hits(hits: pd.DataFrame):
+    return np.count_nonzero(hits.Q.values != NN)
```
May be a bit derivative, but could there be a simple test for this? Or is it already encapsulated well enough by other tests. It's implemented in [enough_valid_hits](https://github.com/gonzaponte/IC/blob/378b6425438f8dbd5acef1a0ad0b8385405b6190/invisible_cities/cities/sophronia.py#L167), which also doesn't seem to be tested explicitly.

jwaiton commented on invisible_cities/reco/hits_functions_test.py:
```diff
@@ -55,54 +77,44 @@ def thresholds(draw, min_value=1, max_value=1):
-    assert (h.Q  > th or h.Q  == NN for h in hits_thresh)
-    assert (h.Qc > th or h.Qc == NN for h in hits_thresh)
+def test_threshold_hits_all_larger_than_th(hits, th, col, on_corrected):
+    hits_thresh = threshold_hits(hits, th, on_corrected = on_corrected)
```
Unrelated; I may open an issue, but it may be worth consolidating the different 'threshold' functions (`cut_over_Q()` in Beersheba, `threshold_hits()` in hit_functions, `hits_thresholder()` in components, `sipm_as_hits()` in Sophronia technically does so as well).

gonzaponte replied:
Agreed. But that's well beyond the scope of this PR. I'll be happy to have an issue os we don't forget.
Of those, `hits_thresholder` is meant to be a component that simply composes existing functions and should not fall in the same category as the other "inner functions" or whatever we want to call them.

jwaiton replied:
Created an issue (#936)

jwaiton commented on invisible_cities/reco/hits_functions.py:
```diff
@@ -1,81 +1,87 @@
 
-def merge_NN_hits(hits : List[evm.Hit], same_peak : bool = True) -> List[evm.Hit]:
+def merge_NN_hits(hits : pd.DataFrame, same_peak : bool = True) -> pd.DataFrame:
     """ Returns a list of the hits where the  energies of NN hits are distributed to the closest hits such that the added energy is proportional to
```
Better docstrings would be nice :smile_cat:

jwaiton commented on invisible_cities/reco/hits_functions.py:
```diff
@@ -1,81 +1,87 @@
+
+        corr_e  = nn_hit.E  * closest.E  / closest.E .sum()
+        corr_ec = nn_hit.Ec * closest.Ec / closest.Ec.sum()
+        corrections.loc[index, "E Ec".split()] += np.stack([corr_e, corr_ec], axis=1)
```
Comments are a bit sparse which in conjunction with the docstring make this a bit harder to comprehend.

For my own sanity (to double check):

The function iterates through all hits with `Q = NN`, finds those within the `same_peak` (line 26) and then for each `nn_hit` row, finds the closest hits that aren't `Q=NN` (lines 29-30) and distributes the `nn_hit` energy across them (lines 33-35). This iterates over each row and is then added to the returned hits (line 39).

gonzaponte replied:
Correct. NN-hits are "merged" with their closest non-NN neighbours.

jwaiton commented on invisible_cities/reco/hits_functions.py:
```diff
@@ -1,81 +1,87 @@
+                            , Qc       = -1
+                            , Ec       = ec
+                            , track_id = -1
+                            , Ep       = -1), index=[0])
```
I can't think of a better place for this to go, but it seems more like a 'type', if that makes sense?

gonzaponte replied:
I understand what you mean. A dataframe with these columns constitutes a sort of hit-collection type. And this function could be an associated method that constructs an empty hit. Sadly, it's not a type just yet. This could be part of a follow-up PR where we find a way of translating specific dataframes into types (I haven't explored if this is in any way possible)

jwaiton commented on invisible_cities/reco/hits_functions.py:
```diff
@@ -1,81 +1,87 @@
+                            , Ep       = -1), index=[0])
+
+
+def apply_threshold(hits : pd.DataFrame, th : float, on_corrected : bool=False) -> pd.DataFrame:
```
No docstrings or comments :)

jwaiton commented on invisible_cities/reco/hits_functions.py:
```diff
@@ -1,81 +1,87 @@
+
+def apply_threshold(hits : pd.DataFrame, th : float, on_corrected : bool=False) -> pd.DataFrame:
+    raw_e_slice = hits.E.sum()
+    cor_e_slice = np.nansum(hits.Ec) + np.finfo(np.float64).eps
```
To aid my understanding, the addition of the epsilon here is to ensure that L78-79 doesn't result in a division by zero, right? Can this issue not also occur with `raw_e_slice`?

jwaiton replied:
Change `np.finfo(np.float64).eps` to `EPSILON`, since its been defined :)

gonzaponte replied:
> To aid my understanding, the addition of the epsilon here is to ensure that L78-79 doesn't result in a division by zero, right?

Yup. Will document.

> Can this issue not also occur with raw_e_slice?

In theory, because of the way we process data, no. But it wouldn't hurt adding a protection, I guess.

> Change np.finfo(np.float64).eps to EPSILON, since its been defined :)

Thanks! I forgot.

jwaiton commented on invisible_cities/reco/hits_functions.py:
```diff
@@ -1,81 +1,87 @@
+    col         = "Qc" if on_corrected else "Q"
+    mask_thresh = hits[col] >= th
+
+    if not mask_thresh.any():
+        first = hits.iloc[0]
+        return empty_hit( first.event, first.time
+                        , first.npeak, first.Xpeak, first.Ypeak
+                        , first.Z, raw_e_slice, cor_e_slice)
```
As far as I understand, this occurs only if there are no hits that pass the threshold. Is this(/will this be) standard behaviour for when a filter is applied to hits/dataframes? It may be beneficial to provide some sort of warning to the user that their cut has removed all hits.

gonzaponte replied:
> Is this(/will this be) standard behaviour for when a filter is applied to hits/dataframes?

Not sure what you mean here. Do we (will we) always produce an empty hit if no sipms survive the threshold? We are doing it now, we could decide to do it differently in the future (slice concatenation could be an option).

As for the warning, it can be added. Not sure it's useful in practice, but we can always produce a wrapper that silences that warning.

jwaiton replied:
Sorry I didn't express myself clearly.

The reason for my concern is due to the experience with `max_num_hits` in `isaura`, where you may end up with isaura not processing any information due to each event exceeding the `max_num_hits`. This process includes no warning (as can be seen [here](https://github.com/next-exp/IC/blob/315ed2e0fd8254db9af44f0815cf1e4da966327b/invisible_cities/cities/components.py#L1329)). 

I see the same pattern of behaviour happening here, as both are 'filters' so to speak. I think it would be generally beneficial for the user to know when particular conditions are met that result in empty outputs (topological, empty hits, or anything in between). 

I also believe this is also beyond the scope of this PR, but it may be nice to consider some sort of generalised solution for this :sweat_smile:.

jwaiton replied:
As discussed offline, the `Filters` group does what I've suggested. There is an issue where the `Filters` tables are empty in isaura files when they shouldn't be (hence my confusion), I've added an issue for this (#939).

On the other hand, with some further reading I've noticed that the `max_num_hits` cut is applied in such a manner within isaura that it would be considered within the `topology_select` table, rather than the `hits_select` table, which I believe should be adjusted. I'll look into this.

jwaiton commented on invisible_cities/reco/hits_functions.py:
```diff
@@ -1,81 +1,87 @@
+                        , first.Z, raw_e_slice, cor_e_slice)
+
+    hits = hits.loc[mask_thresh].copy()
+    qsum = np.nansum(hits.Q) + EPSILON
```
Since `EPSILON` isn't used elsewhere, should it be defined within the function?

EDIT: If it gets used in `apply_threshold()`, this comment can be ignored :)

jwaiton commented on invisible_cities/reco/hits_functions.py:
```diff
@@ -1,81 +1,87 @@
+    hits = hits.loc[mask_thresh].copy()
+    qsum = np.nansum(hits.Q) + EPSILON
+    hits.loc[:, "E" ] = hits.Q / qsum * raw_e_slice
+    hits.loc[:, "Ec"] = hits.Q / qsum * cor_e_slice
```
Similarly in L33-34, I find expressing these sorts of operations without explicit brackets to be quite confusing, but this is just personal preference.

jwaiton commented on invisible_cities/cities/components.py:
```diff
@@ -1540,12 +1540,9 @@ def hits_corrector(map_fname : str, apply_temp : bool) -> Callable:
-        return hitc
+    def correct(hits : pd.DataFrame) -> pd.DataFrame:
+        corr_factors = get_coef(hits.X, hits.Y, hits.Z, hits.time)
+        return hits.assign(Ec = hits.E * corr_factors)
```
Removing the Z correction brings up conflicts with newer PRs right? Specifically #933, will there be an alternative for correcting z position for hits that haven't had z corrections applied?

gonzaponte replied:
Oops, shouldn't have removed the Z correction, actually.

jwaiton commented on invisible_cities/cities/components.py:
```diff
@@ -1497,21 +1497,10 @@ def hits_thresholder(threshold_charge : float, same_peak : bool ) -> Callable:
+    def threshold_hits(hits: pd.DataFrame) -> pd.DataFrame:
+        thr_hits = hif.threshold_hits(    hits, threshold_charge     )
+        mrg_hits = hif.merge_NN_hits (thr_hits, same_peak = same_peak)
+        return mrg_hits
```
Very clean resulting function! I don't see any testing of this component within `components_test.py`, should there be?

jwaiton replied:
Both hit functions `threshold_hits()` and `merge_NN_hits()` are tested, so this isn't an issue.

Would it be possible to try and think of differing names for this function, due to its repetition (`threshold_hits()` calling `hif.threshold_hits()` is a bit ugly).

gonzaponte replied:
The output of a function that simply composes two other functions should be quite predictable. But that's no excuse to avoid testing, so yes, there should always be at least one simple test. I wouldn't stop this PR for that, though, it would be a good exercise for a newcomer.

gonzaponte replied:
Ah and regarding the name, we can rename it. Not because there could be a name collision (its name is only visible within the outer function), but because it can avoid confusion for users.

jwaiton commented on invisible_cities/io/hits_io.py:
```diff
@@ -130,16 +133,10 @@ def load_hits_skipping_NN(DST_file_name : str, group_name : str = 'RECO', table_
-    def write_hits(hits_event):
-        hits_event.store(hits_table)
-    return write_hits
+def hits_writer(hdf5_file, group_name, table_name, *, compression=None):
```
Removing the default `group_name` and `table_name` may cause some confusion when building new cities/modifying the pipe, right? An example where this could be an issue is [here](https://github.com/next-exp/IC/pull/912/files#diff-4241aad994b566cd332688ff48689b507388a77a911e9c9c3a87d73f4b91d99dR1486), although perhaps I'm misunderstanding its implementation.

gonzaponte replied:
Not sure where the confusion comes from. If anything it should make it more clear which table is being written. If I recall correctly, the idea was to use this function in different cities, where each city might write data to a different node.

jwaiton commented on invisible_cities/io/hits_io_test.py:
```diff
@@ -75,32 +79,28 @@ def test_load_hits_double_ratio_e_q_equals_one_skipping_NN(TlMC_hits_skipping_NN
-    assert_allclose(nsipm, dst.nsipm.values)
-    assert_allclose(x    , dst.X    .values)
-    assert_allclose(y    , dst.Y    .values)
-    assert_allclose(np.sqrt(xstd), dst.Xrms .values)
-    assert_allclose(np.sqrt(ystd), dst.Yrms .values)
-    assert_allclose(z    , dst.Z    .values)
-    assert_allclose(q    , dst.Q    .values)
-    assert_allclose(e    , dst.E    .values)
+        write = hits_writer(h5out, "RECO", "Events")
+        write(original_hits)
+
+    read_hits = load_dst(output_file, group = "RECO", node = "Events")
+    assert_dataframes_close(read_hits, original_hits)
```
This is a much cleaner test now! :)

jwaiton commented on invisible_cities/cities/beersheba.py:
```diff
@@ -94,32 +94,8 @@
-    return pd.DataFrame(columns)
-
 def event_info_adder(timestamp : float, dst : pd.DataFrame):
+    dst = dst.astype(dict(npeak=np.float64))
```
Is there a reason that npeak must be a `np.float64` here? Perhaps I'm misunderstanding the purpose of this line.

gonzaponte replied:
Yes, there is... sadly I forgot what the reason was. Hopefully, commenting that line triggers some test failures, which I will use to remember the reason and document this line.

jwaiton replied:
Any update on this? :)

gonzaponte replied:
I think this had something to do with `pandas` doing something wrong. There is some operation (probably `groupby`) that changes the type of the variable. The end result is the same, just of type `float` instead of `int`. After inspection, I've updated the test files to contain ints instead of floats and seems to work, so I'll push it.

jwaiton commented on invisible_cities/cities/components.py:
```diff
@@ -1414,6 +1414,31 @@ def sort_hits(hitc):
     return HitCollection(hitc.event, hitc.time, sorted_hits)
 
 
+def hitc_to_df(hitc: HitCollection):
+    hits = []
+    for hit in hitc.hits:
+        hits.append(pd.DataFrame(dict( event    = hitc.event
+                                     , time     = hitc.time
+                                     , npeak    = hit .npeak
+                                     , Xpeak    = hit .Xpeak
+                                     , Ypeak    = hit .Ypeak
+                                     , nsipm    = hit .nsipm
+                                     , X        = hit .X
+                                     , Y        = hit .Y
+                                     , Xrms     = hit .Xrms
+                                     , Yrms     = hit .Yrms
+                                     , Z        = hit .Z
+                                     , Q        = hit .Q
+                                     , E        = hit .E
+                                     , Qc       = hit .Qc
+                                     , Ec       = hit .Ec
+                                     , track_id = hit .track_id
+                                     , Ep       = hit .Ep), index=[0]))
+    df = pd.concat(hits, ignore_index=True)
+    df = df.astype(dict(event=np.int64, npeak=np.uint16, nsipm=np.uint16, Qc=np.float64, Ec=np.float64, Ep=np.float64))
+    return df
```
I'm assuming this will be removed in a later phase, as well as its usage in the topological functions (`compute_and_write_tracks_info()` for example).

gonzaponte replied:
Yes.

jwaiton commented on invisible_cities/cities/penthesilea.py:
```diff
@@ -101,6 +102,8 @@ def penthesilea( files_in           : OneOrManyFiles
                                    args = ("pmap", "selector_output", "event_number", "timestamp"),
                                    out  = "hits"                                                 )
+
+    to_hits_df            = df.map(hitc_to_df, item="hits")
```
Likewise I assume this is temporary, or are we assuming that cities that rely on hits will retain these conversions functions? If so, It would make sense that both `hitc_to_df()` and `hitc_from_df()` are moved to `hit_functions.py` right? And adequately tested also.

gonzaponte replied:
They will be removed in phase 2.

jwaiton commented on invisible_cities/cities/isaura.py:
```diff
@@ -96,11 +97,17 @@ def isaura( files_in       : OneOrManyFiles
+                                                      , h5out
+                                                      , HitEnergy.E
+                                                      , "hits_select"
+                                                      , write_hits)
```
Moving `write_hits()` into `compute_and_write_tracks_info()` seems to be a bit much for a single function does it not? It's now `compute_and_write_tracks_info_and_hits()`, which I feel could be split up as done previously, unless there is some reason that makes this unfeasible/less elegant.

jwaiton replied:
The obvious issue here is that multiple cities extract topological information, so breaking this function up would inevitably introduce wasted repetition across isaura/esmeralda.

I think a nicer solution would be to implement a topology city that acts on the results regardless of the previous processing (agnostic to beersheba/sophronia/anything else thats given). I may open an issue on this if you think its pertinent.

gonzaponte replied:
> Moving write_hits() into compute_and_write_tracks_info() seems to be a bit much for a single function does it not?

right. That might need a big refactoring, though (haven't delved enough into it). Depending on that I might push it to a later PR, if you are happy with that.

> I think a nicer solution would be to implement a topology city

You probably know better than me. What you say makes sense, but I don't understand if there are any technical limitations. I'm happy to have a topology city like you say.

jwaiton commented on invisible_cities/cities/components.py:
```diff
@@ -1483,7 +1481,9 @@ def compute_and_write_tracks_info(paolina_params, h5out,
     select_and_write_tracks = events_passed_topology.filter, write_tracks
-    write_hits              = (to_hits_df, write_paolina_hits) if write_paolina_hits else None
+
+    to_hits_df = fl.map(hitc_to_df)
+    write_hits = ("paolina_hits", to_hits_df, fl.sink(hits_writer))
```
likewise as mentioned previously. This function now does a lot (extracts and writes topology info, summary, and hits all in one).

jwaiton commented on invisible_cities/io/hits_io_test.py:
```diff
@@ -74,34 +78,32 @@ def test_load_hits_double_ratio_e_q_equals_one_skipping_NN(TlMC_hits_skipping_NN
+    assert_dataframes_close(read_hits, original_hits)
 
 
+# TODO: this test does not test make any sense
```
Assuming this is resolved in phase 2 :smile_cat:

gonzaponte replied:
I think I mean it literally. The test takes data from a file and ensures it belongs to a year range. But the data is static, it doesn't change at any point. It could be modified in `load_hits`, but it doesn't seem to be the case.

jwaiton commented on invisible_cities/reco/hits_functions.py:
```diff
@@ -1,81 +1,87 @@
-                new_hits.append(hit)
-        return new_hits
+    if th <= 0: return hits
+    return (hits.groupby("Z", as_index=False)
```
Very nice refactor :)

jwaiton reviewed (changes requested):
Apologies for the wait! I have done a first read of all the changes. I have undoubtedly missed things, most of which I'll hopefully catch on my next read-through.
Github's review page often hides docstrings if they've not been modified, so I may have commented asking for docstrings where they already exist! Apologies, I tried to fix this but some may have slipped through.

Regardless, docstrings/comments for the new/modified functions are a bit lacking. I'd appreciate it if they could be included. :)

Other than that, good job! I'll have another read-through to ensure that I haven't missed anything soon (in particular related to testing).

gonzaponte committed:
Rewrite fixture to produce hits as dataframes

gonzaponte committed:
Apply Z correction to hits

gonzaponte committed:
Rewrite tests to use dataframes

gonzaponte committed:
Relocate `e_from_q` to `hits_functions`

gonzaponte committed:
Add tests for `e_from_q`

gonzaponte committed:
Factor out SiPM thresholding

gonzaponte committed:
Relocate `sipms_above_threshold` to `hits_functions`

gonzaponte committed:
Add tests for `sipms_above_threshold`

gonzaponte committed:
Add test for `count_valid_hits`

gonzaponte committed:
Improve docs in `hits_functions`

gonzaponte committed:
Clarify usage of EPSILON

gonzaponte committed:
Rename inner function

jwaiton commented on invisible_cities/reco/hits_functions.py:
```diff
@@ -1,81 +1,255 @@
+    # EPSILON added to avoid division by zero
+    # This should be only necessary for Ec, but we apply it to E to be safe
+    hits.loc[:, "E" ] = hits.Q / qsum * (raw_e_slice + EPSILON)
+    hits.loc[:, "Ec"] = hits.Q / qsum * (cor_e_slice + EPSILON)
```
This has been altered, but the underlying issue still remains:
> I find expressing these sorts of operations without explicit brackets to be quite confusing, but this is just personal preference.

I guess what I'm referencing is instead possible replace it with:
`hits.Q / (qsum * (raw_e_slice + EPSILON))`,
although this now looks a bit busy.

gonzaponte replied:
`*` and `/` have the same precedence, therefore a combination of them is evaluated from left to right like:

`(hits.Q / qsum) * (raw_e_slice + EPSILON)`

I'll add the parentheses if you think it's worth it

jwaiton replied:
I think this is just a bias of mine, so feel free to ignore this. 

I like to remove all doubt because I often get it wrong (as I did in my first comment :sweat_smile:).

jacg replied:
FWIW, I'm in favour of the parentheses: they reduce the chance of a human parsing it incorrectly.

Real Programmers (TM) know their precedence tables by heart, never make a mistake in parsing complex expressions and frown upon superfluous (for the compiler) hints. But a fundamental philosophy of IC's design is to cater for real people, not Real Programmers (TM).

jwaiton reviewed (changes requested):
New commits resolve most of the issues, although a few still remain to be addressed (including two from just after the original review, in case they're missed) :)

gonzaponte committed:
Leave `npeak` column as type int

I don't remember the original motivation for it to be a float, probably
some confusion during testing.

gonzaponte committed:
Add parentheses for clarity

gonzaponte committed:
Add some documentation for S12Selector

gonzaponte committed:
Document some components

jwaiton commented on invisible_cities/cities/components.py:
```diff
@@ -934,10 +947,30 @@ def hit_builder(dbfile, run_number, drift_v,
+            generate "empty" (a.k.a. NN) clusters
+        - Assign each cluster the corresponding fraction of the energy in the
+          slice
+    """
```
The documentation is good, but the outer function is missing `typing` type hints (I'm not sure if this matters here). 

The documentation added to `components.py` differs from the 'standard' structure used in most of the city/city related functions. 

This is a consistent pattern I've noticed across IC (the documentation often differs in style and verbosity from file to file). Perhaps I should add an issue addressing this?

jwaiton reviewed (commented):
I have just one relevant comment, and another related to more underlying issues that are highlighted in this PR. Other than these, I'm happy with the PR and will approve once they've been noted :)

gonzaponte committed:
Improve documentation of `hit_builder` and `sipms_as_hits`

jwaiton reviewed (approved):
This PR begins the much desired process of removing the `Hit` class from IC, including changes across many cities and functions to accommodate this removal while retaining the same functionality.

It also adds documentation to some darker areas of the code related to these changes, and has highlighted some new issues (#939, #936).

Approved! Great work, I'm excited for phase 2 :smile_cat:
